test/test_transaction.py

```diff
@@ -22,27 +29,38 @@
 class TestTransaction:  # pylint: disable=too-many-public-methods
     """Unittest for the pymodbus.transaction module."""
 
-    client = None
-    decoder = None
-    _tcp = None
-    _tls = None
-    _rtu = None
-    _ascii = None
-    _manager = None
-    _tm = None
+    client: ModbusBaseSyncClient
+    decoder: ServerDecoder
+    _tcp: ModbusSocketFramer
+    _tls: ModbusTlsFramer
+    _rtu: ModbusRtuFramer
+    _ascii: ModbusAsciiFramer
+    _manager: SyncModbusTransactionManager
 
     # ----------------------------------------------------------------------- #
     # Test Construction
     # ----------------------------------------------------------------------- #
     def setup_method(self):
         """Set up the test environment."""
-        self.client = None
+        self.client = mock.MagicMock(spec=ModbusBaseSyncClient)
+        self.client.framer = mock.MagicMock(spec=ModbusFramer)
         self.decoder = ServerDecoder()
-        self._tcp = ModbusSocketFramer(decoder=self.decoder, client=None)
-        self._tls = ModbusTlsFramer(decoder=self.decoder, client=None)
-        self._rtu = ModbusRtuFramer(decoder=self.decoder, client=None)
-        self._ascii = ModbusAsciiFramer(decoder=self.decoder, client=None)
-        self._manager = SyncModbusTransactionManager(self.client, 0.3, False, False, 3)
+        self._tcp = ModbusSocketFramer(
+            decoder=self.decoder, client=ModbusTcpClient("127.0.0.1")
```

Comment:
why initialize a client here?  I don't see how it improves anything over `client = None`

Reply:
> why initialize a client here? I don't see how it improves anything over `client = None`

None isn't an allowable type for [`client: ModbusBaseSyncClient`](https://github.com/pymodbus-dev/pymodbus/blob/978ad88b2eb7482d681bbfcef46a8f0f05547f84/pymodbus/transaction.py#L137). Not sure if it's better to use a real client or a mock client here.

Reply:
I changed this to use `mock.create_autospec` for creating mock client objects.

Reply:
> > why initialize a client here? I don't see how it improves anything over `client = None`
> 
> `None` isn't an allowable type for [`client: ModbusBaseSyncClient`](https://github.com/pymodbus-dev/pymodbus/blob/978ad88b2eb7482d681bbfcef46a8f0f05547f84/pymodbus/transaction.py#L137). Not sure if it's better to use a real client or a mock client here.

Actually, only `_manager` needs a client (mock or real).  The others all accept a default argument of `client = None`